### ISSUE TITLE
[java] UnusedPrivateField doesn't find annotated unused private fields anymore

### ISSUE THREAD

**Issue by phoenix384** [pmd/pmd#4166]
**Rule**: [UnusedPrivateField](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedprivatefield)

With PMD 6.50 the rule UnusedPrivateField was changed to completely ignore any private field with an annotation (see #4100).
But what if the previous behaviour was totally ok? How to restore it? In my case I do want to find unused private fields even with annotation.

If you need a use case for that: Java Selenium Tests
There you annotate WebElements with the locator that points to the corresponding Element on the web page.

It looks like this:

```java
@FindBy(id = "whateverID")
private WebElement myElement;
```

A class like this could have a bunch of WebElements and if some of them are unused, you don't see it.
If I don't use myElement, I want to remove it and I want PMD to find it.

Maybe at least a kind of whitelist could be added to configure PMD to find fields even if they have annotations that are on this list.

**Comment by adangel** [comment]
Thanks for sharing your use case.

As written already in the comment https://github.com/pmd/pmd/pull/4100#issuecomment-1263264463:

> I decided to deprecate the property "ignoreAnnotations". If there are good examples for false negatives, we might think of adding a new property to change the behavior, to deal with some annotations in a special way (e.g. "reportForAnnotations") or have a boolean flag (e.g. "reportRegardlessOfAnnotations"=true) which could be enabled if needed and would report all unused private fields whether annotated or not.

We can easily add either "reportForAnnotations" or "reportRegardlessOfAnnotations". What I want to avoid however is, that we start maintaining now a whitelist for Java frameworks in PMD (as we did before with the blacklist).

Would that work for you - given the rule has such a property - that you configure the rule to your needs?
Would you rather have a simple flag ("reportRegardlessOfAnnotations") or a more fine granular property ("reportForAnnotations") where you can list individual annotation classes?

**Comment by adangel** [comment]
I'll add a new property `annotations` (like in https://pmd.github.io/latest/pmd_rules_java_errorprone.html#missingstaticmethodinnoninstantiatableclass) which can be configured for this use case.
Maybe we can use `org.openqa.selenium.support.FindBy` as a default value.

**Comment by phoenix384** [comment]
Thank you. That works.

**Comment by Baklanozavr** [comment]
Greetings!

The `UnusedPrivateField` and `UnusedPrivateMethod` rules currently have inconsistent annotation properties, despite their similar goals. While `UnusedPrivateMethod` uses `ignoredAnnotations`, `UnusedPrivateField` relies on `ignoredFieldNames` and `reportForAnnotations`. This difference makes configuration confusing.

Both rules could benefit from having both `reportForAnnotations` and `ignoredAnnotations` properties. This would allow users to choose their preferred behaviour: either do or don't ignore annotated members. Managing a `reportForAnnotations` list can become cumbersome for complex projects (the list exceeded 40 items in my case).

Moreover, the current logic around `reportForAnnotations` is fragile. For example, I can imagine a situation with a custom annotation for code ownership over every class in a project. When this annotation is renamed, the `reportForAnnotations` list might not be updated, which will lead to missed unused fields.

Consistent annotation handling across these rules would improve usability.

**Comment by oowekyala** [comment]
Hello, thanks for your feedback. Can you please open a new issue?

**Comment by Baklanozavr** [comment]
@oowekyala Thank you for the swift reply! New issue created: https://github.com/pmd/pmd/issues/5808
